src/shared/awsagent/roles.go

```diff
@@ -59,25 +71,15 @@ func (a *Agent) initProfileCache(ctx context.Context) (err error) {
 		}
 	}()
 	a.profileCacheOnce.Do(func() {
-		var nextToken *string = nil
-		output, listProfileErr := a.rolesAnywhereClient.ListProfiles(ctx, &rolesanywhere.ListProfilesInput{NextToken: nextToken})
-		if listProfileErr != nil {
-			err = errors.Errorf("failed to list profiles: %w", listProfileErr)
+		var nextToken *string
+		nextToken, err = a.loadNextProfileCachePage(ctx, nil)
+		if err != nil {
 			return
 		}
 
-		for _, profile := range output.Profiles {
-			a.profileNameToId[*profile.Name] = *profile.ProfileId
-		}
-
-		for output.NextToken != nil {
-			nextToken = output.NextToken
-			for _, profile := range output.Profiles {
-				a.profileNameToId[*profile.Name] = *profile.ProfileId
-			}
```

Comment:
Note for reviewer: the bug was that this line was not called on the last chunk of the iteration (when `nextToken = nil` again).